### PR TITLE
got premium users to upgrade, on to the downgrade

### PR DESCRIPTION
**Pull request by AbdullahAlger**

### CODE REVIEW

**Comment by AbdullahAlger** [comment]
@eliotsykes 
I am not really happy with this. I somehow think I need another model for the downgrade. It's not a downgrade for the charge, but for the user. Any ideas how it could be better?

File: app/models/user.rb [hunk 1]
```diff
@@ -22,8 +17,29 @@ def premium?
     role == 'premium'
   end
 
+  def standard?
+    role == 'standard'
+  end
+
   def upgrade_account
-    self.update_attributes(role: "premium")
+    if self.role == "standard"
```
Comment (eliotsykes) [inline app/models/user.rb]: Consider using the `standard?` method on line 25.

File: app/models/user.rb [hunk 2]
```diff
@@ -22,8 +17,29 @@ def premium?
...
+    end
+  end
+
+  def can_privatize_wiki?(wiki)
+    (self.premium? && wiki.is_owned_by?(self)) || self.admin?
+  end
+
+  def make_wikis_public
+    wikis.each do |wiki|
+      wiki.update_attributes(private: false)
```
Comment (eliotsykes) [inline app/models/user.rb]: Consider introducing a `make_public` instance method on `Wiki` and call it on line 36.

File: db/seeds.rb [hunk 3]
```diff
@@ -6,6 +6,17 @@
 #   cities = City.create([{ name: 'Chicago' }, { name: 'Copenhagen' }])
 #   Mayor.create(name: 'Emanuel', city: cities.first)
 
+
+1.times do
```
Comment (eliotsykes) [inline db/seeds.rb]: Consider if the `1.times` block is doing something useful.

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      expect(user.valid?).to eq false
+      expect(user.errors[:role]).to include "can't be blank"
+    end
+
+    it "should have a valid role" do
+      user = User.new(role: "foo")
+      expect(user.valid?).to eq false
+      expect(user.errors[:role]).to include("admin")
+    end
+
```
Comment (eliotsykes) [inline spec/models/user_spec.rb]: Consider adding test coverage for `admin`, `standard`, and `premium` being valid roles.

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    it "should validate whether a user has a role" do
+      user = User.new(role: "")
+      expect(user.valid?).to eq false
+      expect(user.errors[:role]).to include "can't be blank"
+    end
+
+    it "should have a valid role" do
+      user = User.new(role: "foo")
+      expect(user.valid?).to eq false
+      expect(user.errors[:role]).to include("admin")
```
Comment (eliotsykes) [inline spec/models/user_spec.rb]: Line 25 expectation isn't clear in what its purpose is. In addition the error message substring "admin" could (one-day) be in multiple validation error messages so this isn't proving the right validation error is being triggered. Consider including the full error message on line 25.

**Comment by eliotsykes** [comment]
> I am not really happy with this. I somehow think I need another model for the downgrade. It's not a downgrade for the charge, but for the user. Any ideas how it could be better?

You're right this doesn't feel like the responsibility of the `ChargesController`. Try moving the `downgrade` action to `UsersController`. There's no need to add a new model dedicated to downgrades at this point but its a good thought. 

In the future, if downgrades need to be tracked (e.g. accounting department want to generate reports on dates downgrades occur, how much money is refunded, etc.), then a dedicated `Downgrade` or `SubscriptionChange` or similar model would be A+.

File: config/routes.rb [hunk 6]
```diff
@@ -4,6 +4,10 @@
   root to: "home#index"
 
   devise_for :users
+
+  devise_scope :user do
+    get "/downgrade" => "charges#downgrade"
```
Comment (eliotsykes) [inline config/routes.rb]: HTTP GETs don't usually change data on a model. Review the available HTTP verbs and choose one that feels the most appropriate for modifying a user with a downgrade.

File: app/models/user.rb (hunk 2 above)
Comment (AbdullahAlger) [inline app/models/user.rb]: done

**Pull request merged**